File: vi/codes/dart/chapter_dynamic_programming/knapsack.dart
/**
 * File: knapsack.dart
 * Created Time: 2023-08-11
 * Author: liuyuxin (gvenusleo@gmail.com)
 */

import 'dart:math';

/* Cái túi 0-1: Tìm kiếm vét cạn */
int knapsackDFS(List<int> wgt, List<int> val, int i, int c) {
  // Nếu đã chọn hết vật phẩm hoặc túi không còn dung lượng, trả về giá trị 0
  if (i == 0 || c == 0) {
    return 0;
  }
  // Nếu vượt quá dung lượng túi, chỉ có thể chọn không đặt vào
  if (wgt[i - 1] > c) {
    return knapsackDFS(wgt, val, i - 1, c);
  }
  // Tính giá trị lớn nhất của việc không đặt vào và đặt vào vật phẩm i
  int no = knapsackDFS(wgt, val, i - 1, c);
  int yes = knapsackDFS(wgt, val, i - 1, c - wgt[i - 1]) + val[i - 1];
  // Trả về giá trị lớn hơn trong hai lựa chọn
  return max(no, yes);
}

/* Cái túi 0-1: Tìm kiếm ghi nhớ */
int knapsackDFSMem(
  List<int> wgt,
  List<int> val,
  List<List<int>> mem,
  int i,
  int c,
) {
  // Nếu đã chọn hết vật phẩm hoặc túi không còn dung lượng, trả về giá trị 0
  if (i == 0 || c == 0) {
    return 0;
  }
  // Nếu đã có bản ghi, trả về ngay
  if (mem[i][c] != -1) {
    return mem[i][c];
  }
  // Nếu vượt quá dung lượng túi, chỉ có thể chọn không đặt vào
  if (wgt[i - 1] > c) {
    return knapsackDFSMem(wgt, val, mem, i - 1, c);
  }
  // Tính giá trị lớn nhất của việc không đặt vào và đặt vào vật phẩm i
  int no = knapsackDFSMem(wgt, val, mem, i - 1, c);
  int yes = knapsackDFSMem(wgt, val, mem, i - 1, c - wgt[i - 1]) + val[i - 1];
  // Ghi lại và trả về giá trị lớn hơn trong hai lựa chọn
  mem[i][c] = max(no, yes);
  return mem[i][c];
}

/* Cái túi 0-1: Quy hoạch động */
int knapsackDP(List<int> wgt, List<int> val, int cap) {
  int n = wgt.length;
  // Khởi tạo bảng dp
  List<List<int>> dp = List.generate(n + 1, (index) => List.filled(cap + 1, 0));
  // Chuyển trạng thái
  for (int i = 1; i <= n; i++) {
    for (int c = 1; c <= cap; c++) {
      if (wgt[i - 1] > c) {
        // Nếu vượt quá dung lượng túi, không chọn vật phẩm i
        dp[i][c] = dp[i - 1][c];
      } else {
        // Giá trị lớn hơn giữa việc không chọn và chọn vật phẩm i
        dp[i][c] = max(dp[i - 1][c], dp[i - 1][c - wgt[i - 1]] + val[i - 1]);
      }
    }
  }
  return dp[n][cap];
}

/* Cái túi 0-1: Quy hoạch động tối ưu không gian */
int knapsackDPComp(List<int> wgt, List<int> val, int cap) {
  int n = wgt.length;
  // Khởi tạo bảng dp
  List<int> dp = List.filled(cap + 1, 0);
  // Chuyển trạng thái
  for (int i = 1; i <= n; i++) {
    // Duyệt theo thứ tự ngược
    for (int c = cap; c >= 1; c--) {
      if (wgt[i - 1] <= c) {
        // Giá trị lớn hơn giữa việc không chọn và chọn vật phẩm i
        dp[c] = max(dp[c], dp[c - wgt[i - 1]] + val[i - 1]);
      }
    }
  }
  return dp[cap];
}

/* Driver Code */
void main() {
  List<int> wgt = [10, 20, 30, 40, 50];
  List<int> val = [50, 120, 150, 210, 240];
  int cap = 50;
  int n = wgt.length;

  // Tìm kiếm vét cạn
  int res = knapsackDFS(wgt, val, n, cap);
  print("Maximum item value not exceeding knapsack capacity is $res");

  // Tìm kiếm ghi nhớ
  List<List<int>> mem =
      List.generate(n + 1, (index) => List.filled(cap + 1, -1));
  res = knapsackDFSMem(wgt, val, mem, n, cap);
  print("Maximum item value not exceeding knapsack capacity is $res");

  // Quy hoạch động
  res = knapsackDP(wgt, val, cap);
  print("Maximum item value not exceeding knapsack capacity is $res");

  // Quy hoạch động tối ưu không gian
  res = knapsackDPComp(wgt, val, cap);
  print("Maximum item value not exceeding knapsack capacity is $res");
}

File: vi/codes/ruby/chapter_dynamic_programming/knapsack.rb
=begin
File: knapsack.rb
Created Time: 2024-05-29
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

### Bài toán cái túi 0-1: tìm kiếm vét cạn ###
def knapsack_dfs(wgt, val, i, c)
  # Nếu đã chọn hết vật phẩm hoặc túi không còn dung lượng, trả về giá trị 0
  return 0 if i == 0 || c == 0
  # Nếu vượt quá dung lượng túi, chỉ có thể chọn không cho vào
  return knapsack_dfs(wgt, val, i - 1, c) if wgt[i - 1] > c
  # Tính giá trị lớn nhất giữa việc không cho vào và cho vật phẩm i vào
  no = knapsack_dfs(wgt, val, i - 1, c)
  yes = knapsack_dfs(wgt, val, i - 1, c - wgt[i - 1]) + val[i - 1]
  # Trả về giá trị lớn hơn trong hai lựa chọn
  [no, yes].max
end

### Bài toán cái túi 0-1: tìm kiếm ghi nhớ ###
def knapsack_dfs_mem(wgt, val, mem, i, c)
  # Nếu đã chọn hết vật phẩm hoặc túi không còn dung lượng, trả về giá trị 0
  return 0 if i == 0 || c == 0
  # Nếu đã có bản ghi, trả về ngay
  return mem[i][c] if mem[i][c] != -1
  # Nếu vượt quá dung lượng túi, chỉ có thể chọn không cho vào
  return knapsack_dfs_mem(wgt, val, mem, i - 1, c) if wgt[i - 1] > c
  # Tính giá trị lớn nhất giữa việc không cho vào và cho vật phẩm i vào
  no = knapsack_dfs_mem(wgt, val, mem, i - 1, c)
  yes = knapsack_dfs_mem(wgt, val, mem, i - 1, c - wgt[i - 1]) + val[i - 1]
  # Ghi lại và trả về giá trị lớn hơn trong hai lựa chọn
  mem[i][c] = [no, yes].max
end

### Bài toán cái túi 0-1: quy hoạch động ###
def knapsack_dp(wgt, val, cap)
  n = wgt.length
  # Khởi tạo bảng dp
  dp = Array.new(n + 1) { Array.new(cap + 1, 0) }
  # Chuyển trạng thái
  for i in 1...(n + 1)
    for c in 1...(cap + 1)
      if wgt[i - 1] > c
        # Nếu vượt quá dung lượng túi, không chọn vật phẩm i
        dp[i][c] = dp[i - 1][c]
      else
        # Giá trị lớn hơn giữa việc không chọn và chọn vật phẩm i
        dp[i][c] = [dp[i - 1][c], dp[i - 1][c - wgt[i - 1]] + val[i - 1]].max
      end
    end
  end
  dp[n][cap]
end

### Bài toán cái túi 0-1: quy hoạch động tối ưu không gian ###
def knapsack_dp_comp(wgt, val, cap)
  n = wgt.length
  # Khởi tạo bảng dp
  dp = Array.new(cap + 1, 0)
  # Chuyển trạng thái
  for i in 1...(n + 1)
    # Duyệt theo thứ tự ngược
    for c in cap.downto(1)
      if wgt[i - 1] > c
        # Nếu vượt quá dung lượng túi, không chọn vật phẩm i
        dp[c] = dp[c]
      else
        # Giá trị lớn hơn giữa việc không chọn và chọn vật phẩm i
        dp[c] = [dp[c], dp[c - wgt[i - 1]] + val[i - 1]].max
      end
    end
  end
  dp[cap]
end

### Mã chương trình chính ###
if __FILE__ == $0
  wgt = [10, 20, 30, 40, 50]
  val = [50, 120, 150, 210, 240]
  cap = 50
  n = wgt.length

  # Tìm kiếm vét cạn
  res = knapsack_dfs(wgt, val, n, cap)
  puts "Maximum item value not exceeding knapsack capacity is #{res}"

  # Tìm kiếm ghi nhớ
  mem = Array.new(n + 1) { Array.new(cap + 1, -1) }
  res = knapsack_dfs_mem(wgt, val, mem, n, cap)
  puts "Maximum item value not exceeding knapsack capacity is #{res}"

  # Quy hoạch động
  res = knapsack_dp(wgt, val, cap)
  puts "Maximum item value not exceeding knapsack capacity is #{res}"

  # Quy hoạch động tối ưu không gian
  res = knapsack_dp_comp(wgt, val, cap)
  puts "Maximum item value not exceeding knapsack capacity is #{res}"
end
